go/libraries/doltcore/sqle/globalstate/auto_increment_tracker.go

```diff
@@ -110,20 +106,90 @@ func (a AutoIncrementTracker) Next(tbl string, insertVal interface{}) (uint64, e
 	return given, nil
 }
 
+// CoerceAutoIncrementValue converts |val| into an AUTO_INCREMENT sequence value
+func CoerceAutoIncrementValue(val interface{}) (uint64, error) {
+	switch typ := val.(type) {
+	case float32:
+		val = math.Round(float64(typ))
+	case float64:
+		val = math.Round(typ)
+	}
+
+	var err error
+	val, err = sql.Uint64.Convert(val)
+	if err != nil {
+		return 0, err
+	}
+	if val == nil || val == uint64(0) {
+		return 0, nil
+	}
+	return val.(uint64), nil
+}
+
+// Set sets the auto increment value for the table named, if it's greater than the one already registered for this
+// table. Otherwise, the update is silently disregarded. So far this matches the MySQL behavior, but Dolt uses the
+// maximum value for this table across all branches.
 func (a AutoIncrementTracker) Set(tableName string, val uint64) {
 	a.mu.Lock()
 	defer a.mu.Unlock()
-	a.sequences[tableName] = val
+
+	tableName = strings.ToLower(tableName)
+
+	existing := a.sequences[tableName]
+	if val > existing {
+		a.sequences[strings.ToLower(tableName)] = val
+	}
 }
 
+// AddNewTable initializes a new table with an auto increment column to the tracker, as necessary
 func (a AutoIncrementTracker) AddNewTable(tableName string) {
 	a.mu.Lock()
 	defer a.mu.Unlock()
-	a.sequences[tableName] = uint64(1)
+
+	tableName = strings.ToLower(tableName)
+	// only initialize the sequence for this table if no other branch has such a table
+	if _, ok := a.sequences[tableName]; !ok {
+		a.sequences[tableName] = uint64(1)
+	}
 }
 
-func (a AutoIncrementTracker) DropTable(tableName string) {
+// DropTable drops the table with the name given.
+// To establish the new auto increment value, callers must also pass all other working sets in scope that may include
+// a table with the same name, omitting the working set that just deleted the table named.
+func (a AutoIncrementTracker) DropTable(ctx context.Context, tableName string, wses ...*doltdb.WorkingSet) error {
```

Comment:
this seems quite racy. state for concurrent transactions may be held only in this tracker. Consider transactions that  are concurrently doing a `SET AUTO_INCREMENT = x` and a `TRUNCATE`. 
Are we sure we need to update the tracker at all in the case of a `TRUNCATE` or drop table? MySQL compliance would dictate that we do for `TRUNCATE`, but we're already fudging that with the cross-branch sequence communication.

Reply:
It's not really racy, but it's kinda busted in the general sense (not actually transactional). I'm not super worried about this though because the way this works in MySQL is already very strange, and MySQL doesn't respect transaction boundaries for DDL statements. If it's a problem we can just vary slightly farther from MySQL and not adjust auto increment values on drop table.